### PR TITLE
Add minimal support for RZ/G3S SMARC (CM33)

### PR DESCRIPTION
**Pull request by binhnguyen2434**
- Add minimal support for RZ/G3S SMARC as following:

Interface | Controller | Driver/Component
-- | -- | --
GPIO | on-chip | gpio
Pinctrl | on-chip | gpio
UART | on-chip | serial

Need: https://github.com/zephyrproject-rtos/hal_renesas/pull/50

docs: https://builds.zephyrproject.io/zephyr/pr/82016/docs/boards/renesas/rzg3s_smarc/doc/index.html

### CODE REVIEW

**Review by nordicjm — CHANGES_REQUESTED** [review]
drivers must correctly select PINCTRL

File: boards/renesas/rzg3s_smarc/board.cmake [hunk 1]
```diff
@@ -0,0 +1,9 @@
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+board_runner_args(jlink "--device=R9A08G045S33_M33_0" "--speed=15000")
+include(${ZEPHYR_BASE}/boards/common/jlink.board.cmake)
+
+set_property(GLOBAL APPEND PROPERTY extra_post_build_commands
+    COMMAND ${CMAKE_OBJCOPY} -I binary -O srec --srec-forceS --adjust-vma 0x23000 ${ZEPHYR_BINARY_DIR}/${KERNEL_NAME}.bin ${ZEPHYR_BINARY_DIR}/${KERNEL_NAME}.srec
+)
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/board.cmake:9]: also quite puzzled by this, what is this doing? Why does this need to happen at a board level? Why can't this be done properly by the build system?

Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/board.cmake:9]: Thank you for your feedback.
This will add a custom post-build step to convert the kernel binary .bin file to an S-record .srec file.
As RZ/G3S boot sequence, srec file is required for flashing. 
We plan to support more RZ boards soon, the boot sequence of other RZ boards would be different and may not require srec file so this custom post-build step shall be defined in board level.

File: soc/renesas/rz/rzg3s/Kconfig.soc [hunk 2]
```diff
@@ -0,0 +1,20 @@
...
+	  Renesas RZ/G3S series
+
+config SOC_SERIES
+	default "rzg3s" if SOC_SERIES_RZG3S
+
+config SOC_R9A08G045S33GBG
+	bool
+	select SOC_SERIES_RZG3S
+	help
+	  R9A08G045S33GBG
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply soc/renesas/rz/rzg3s/Kconfig.soc:17]: Thank you for the feedback.
I updated as your suggestion.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 3]
```diff
@@ -0,0 +1,253 @@
+.. _rzg3s_smarc:
+
+#################################
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: Thank you for the feedback.
I deleted the redundant line.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 4]
```diff
@@ -0,0 +1,253 @@
...
+RZ/G3S SMARC Evaluation Board Kit
+#################################
+
+.. _Overview:
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: Thank you for the feedback.
I deleted the redundant lines.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 5]
```diff
@@ -0,0 +1,253 @@
...
+	:width: 600px
+	:align: center
+	:alt: RZ/G3S group feature
+
+	RZ/G3S block diagram (Credit: Renesas Electronics Corporation)
+
+Supported Features
+==================
+
+The Renesas ``rzg3s_smarc`` board configuration supports for ARM Cortex-M33 System Core without FPU
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: Thank you for the feedback.
I updated as your suggestion.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig [hunk 6]
```diff
@@ -0,0 +1,22 @@
...
+
+CONFIG_SYS_CLOCK_EXISTS=y
+
+CONFIG_XIP=n
+
+CONFIG_MAIN_STACK_SIZE=4096
+
+# Enable PINCTRL driver
+CONFIG_PINCTRL=y
+
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig]: Thank you for the feedback.
I updated to select PINCTRL in serial driver and removed the enable PINCTRL in board defconfig.
Please kindly check it.

File: drivers/gpio/gpio_renesas_rz.c [hunk 7]
```diff
@@ -0,0 +1,496 @@
...
+		/* Disconnect mode */
+		ioport_config_data = 0;
+	} else if (!(flags & GPIO_OPEN_DRAIN)) {
+		/* PM register */
+		if (flags & GPIO_INPUT) {
+			if (flags & GPIO_OUTPUT) {
+				ioport_config_data &= GPIO_RZ_PIN_CONFIGURE_INPUT_OUTPUT_RESET;
+				ioport_config_data |= IOPORT_CFG_PORT_DIRECTION_OUTPUT_INPUT;
+			} else {
+				ioport_config_data &= ~(0x3 << 2);
```
Comment (tgorochowik) [inline drivers/gpio/gpio_renesas_rz.c:119]: Can we use a dedicated define here as well?

File: drivers/gpio/gpio_renesas_rz.c [hunk 8]
```diff
@@ -0,0 +1,496 @@
...
+	volatile uint32_t *tssr = &R_INTC_IM33->TSSR0;
+	volatile uint32_t *titsr = &R_INTC_IM33->TITSR0;
+	volatile uint32_t *tscr = &R_INTC_IM33->TSCR;
+
+	tssr = &tssr[tint_num / 4];
+	titsr = &titsr[tint_num / 16];
+
+	irq_disable(GPIO_RZ_TINT_IRQ_GET(tint_num));
+	*tssr &= ~(0xFF << GPIO_RZ_TSSR_OFFSET(tint_num));
+	*titsr &= ~(0x3 << GPIO_RZ_TITSR_OFFSET(tint_num));
```
Comment (tgorochowik) [inline drivers/gpio/gpio_renesas_rz.c:270]: Can we explain this mask? Either through a define or at least a dedicated comment

**Review by tgorochowik — COMMENTED** [review]
Thank you for the contribution! Some initial minor comments

File: drivers/gpio/gpio_renesas_rz.c (hunk 7 above)
Comment (binhnguyen2434) [reply drivers/gpio/gpio_renesas_rz.c]: Thank you for the feedback.
I changed to the dedicated defined macro.
Please kindly check it.

File: drivers/gpio/gpio_renesas_rz.c (hunk 8 above)
Comment (binhnguyen2434) [reply drivers/gpio/gpio_renesas_rz.c:272]: Thank you for the feedback.
This mask is used to get to position of the bit related to interrupt number in the register and clear it to default value.
I also added the comment it.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/board.cmake (hunk 1 above)
Comment (tejlmand) [reply boards/renesas/rzg3s_smarc/board.cmake:9]: but Zephyr already support conversion to srec in a way which also uses the toolchain infrastructure meaning that you're not relying on a specific toolchain.

https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/CMakeLists.txt#L1815-L1835

All you need to do is set `CONFIG_BUILD_OUTPUT_S19=y`.
Ref:
https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/Kconfig.zephyr#L749

So why are you not using existing functionality ?

**Review by tejlmand — CHANGES_REQUESTED** [review]
Please use existing srec / s19 conversion feature in Zephyr, see: https://github.com/zephyrproject-rtos/zephyr/pull/82016#discussion_r1865407606

File: boards/renesas/rzg3s_smarc/board.cmake (hunk 1 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/board.cmake:9]: Thank you for providing this information.
We tried to use it, but it only generates `*.s19` file. The bootloader only support srec file format.
Are there any other options or configurations to enable srec generation within the build system?
If no, we should consider updating [CMakeLists.txt](https://github.com/zephyrproject-rtos/zephyr/blob/6f26952828b037e46934a5b98d7d1d05eab45aaf/CMakeLists.txt#L1815-L1835) to support srec generation.

Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/board.cmake]: Sorry, I misunderstood that s19 is a different format from srec.
I updated as your suggestion. 
Please kindly check it.

File: west.yml [hunk 9]
```diff
@@ -214,7 +214,7 @@ manifest:
         - hal
     - name: hal_renesas
       path: modules/hal/renesas
-      revision: 10326518701e25bf336a2eaeb8b5820110e4e6a3
+      revision: ca0be9e9c56bdedf2f18426639c5aca79a4e06ae
```
Comment (nordicjm) [inline west.yml:217]: this commit should be before the things that need/use it

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 10]
```diff
@@ -0,0 +1,250 @@
...
+	:width: 600px
+	:align: center
+	:alt: RZ/G3S group feature
+
+	RZ/G3S block diagram (Credit: Renesas Electronics Corporation)
+
+Supported Features
+==================
+
+The Renesas ``rzg3s_smarc/r9a08g045s33gbg/cm33`` board target supports for ARM Cortex-M33 System Core without FPU
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/doc/index.rst:61]:
```suggestion
The ``rzg3s_smarc/r9a08g045s33gbg/cm33`` board target supports ARM Cortex-M33 System Core without FPU
```

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 11]
```diff
@@ -0,0 +1,250 @@
...
++-----------+------------+-------------------------------------+
+| UART      | on-chip    | serial                              |
++-----------+------------+-------------------------------------+
+
+Other hardware features are currently not supported by the port.
+
+Programming and Debugging
+*************************
+
+RZ/G3S-EVKIT is designed to start different systems on cores.
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/doc/index.rst:83]:
```suggestion
RZ/G3S-EVKIT is designed to start different systems on different cores.
```
(I think?)

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 12]
```diff
@@ -0,0 +1,250 @@
...
+=========
+
+It is possible to load and execute a Zephyr application binary on
+this board on the Cortex-M33 System Core from
+the internal SRAM, using ``JLink`` debugger (:ref:`jlink-debug-host-tools`).
+
+.. note::
+
+    Currently it's required Renesas BL2 TF-A to be started on Cortex-A55 System Core
+    before starting Zephyr. As BL2 TF-A configures clocks and Cortex-M33 System Core before starting it.
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/doc/index.rst:146]:
```suggestion
    Currently it's required Renesas BL2 TF-A to be started on Cortex-A55 System Core
    before starting Zephyr as it configures clocks and the Cortex-M33 System Core before starting it.
```

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 13]
```diff
@@ -0,0 +1,250 @@
...
+.. zephyr-app-commands::
+   :zephyr-app: samples/hello_world
+   :board: rzg3s_smarc/r9a08g045s33gbg/cm33
+   :goals: build debug
+
+Flashing
+========
+
+Zephyr application can be flashed to QSPI storage and then loaded by
+Renesas BL2 TF-A running on Cortex-A55 System Core and staring binary at Cortex-M33 System Core.
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/doc/index.rst:159]:
```suggestion
Renesas BL2 TF-A running on the Cortex-A55 System Core and starting binary on the Cortex-M33 System Core.
```

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 14]
```diff
@@ -0,0 +1,250 @@
...
+
+The Zephyr application binary has to be converted to Motorolla S-record `SREC`_ format
+which is generated automatically in Zephyr application build directory (zephyr.s19).
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/doc/index.rst:162]:
```suggestion
which is generated automatically in Zephyr application build directory.
```
the name is determined by a Kconfig

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig [hunk 15]
```diff
@@ -0,0 +1,21 @@
+# Copyright (c) 2024 EPAM Systems
+# Copyright (c) 2024 Renesas Electronics Corporation
+# SPDX-License-Identifier: Apache-2.0
+
+CONFIG_SYS_CLOCK_EXISTS=y
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig:5]: should be set at soc level?

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig [hunk 16]
```diff
@@ -0,0 +1,21 @@
...
+
+CONFIG_XIP=n
+
+CONFIG_MAIN_STACK_SIZE=4096
+
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig:10]:
```suggestion
```

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig [hunk 17]
```diff
@@ -0,0 +1,21 @@
...
+CONFIG_SERIAL=y
+
+# Enable console
+CONFIG_CONSOLE=y
+CONFIG_UART_CONSOLE=y
+
+# This is required to support debug with xSPI boot when Zephyr already booted by TF-A
+CONFIG_INIT_ARCH_HW_AT_BOOT=y
+
+CONFIG_BUILD_OUTPUT_S19=y
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig:21]: should be set at soc level

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33.dts [hunk 18]
```diff
@@ -23,6 +24,34 @@
...
+		sw0 = &sw_1;
+		sw1 = &sw_2;
+		sw2 = &sw_3;
+	};
+
+	buttons {
+		compatible = "gpio-keys";
+
+		sw_1: button_1 {
+			gpios = <&gpio18 0 GPIO_ACTIVE_LOW>;
```
Comment (nordicjm) [inline boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33.dts:37]: either move GPIO driver commit before board and move the board changes in this commit into the board commit, or move the board changes here into a separate commit

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 19]
```diff
@@ -0,0 +1,250 @@
...
+* Input when asked:
+
+.. code-block:: console
+
+    ===== Please Input Program Top Address ============
+      Please Input : H'23000
+    ===== Please Input Qspi Save Address ===
+      Please Input : H'200000
+
+* Then send Zephyr **srec** file from terminal (use ''ascii'' mode)
```
Comment (tejlmand) [inline boards/renesas/rzg3s_smarc/doc/index.rst:184]: perhaps inform the reader that the file with have the `s19` extension, to minimize the risk of srec / s19 confusion you experienced yourself.

Non-blocking comment.

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig (hunk 16 above)
Comment (tejlmand) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig:10]: I agree 4096 looks a bit high.
@binhnguyen2434 can you explain why so much main stack is needed ?

**Review by tejlmand — DISMISSED** [review]
please address comments by @nordicjm.

Rest lgtm, so +1.

Thanks for the contribution.

File: west.yml (hunk 9 above)
Comment (binhnguyen2434) [reply west.yml:218]: I moved this commit to the top.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 10 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: I updated as your suggestion.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 11 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: That's sound better. Thank you.
I updated as your suggestion.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 13 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: I updated as your suggestion.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 14 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: I removed the fixed name and mentioned the file extension s19.

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig (hunk 15 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig]: Thank you.
I moved it to `soc\renesas\rz\rzg3s\Kconfig.defconfig`

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig (hunk 16 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig]: I verified with default stack 1024 of system. That's work fine,
So, this config is unnecessary, I removed it.

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig (hunk 17 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33_defconfig]: Thank you.
I moved them to `soc\renesas\rz\rzg3s\Kconfig.defconfig`

File: boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33.dts (hunk 18 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/rzg3s_smarc_r9a08g045s33gbg_cm33.dts:37]: Thank you.
I moved this modification of GPIO to the board commit.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 19 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: That's right. It is better to change it to s19 file. Updated.
`* Then send Zephyr **s19** file from terminal (use ''ascii'' mode)`

**Review by nordicjm — DISMISSED** [review]

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 20]
```diff
@@ -0,0 +1,250 @@
...
+   :goals: build flash
+   :compact:
+
+References
+**********
+- `RZ/G3S Group Website`_
+- `RZG3S-EVKIT Website`_
+- `RZ/G3S SMARC Module Board User's Manual Hardware`_
+- `RZ SMARC Series Carrier Board II User's Manual Hardware`_
+- `SMARC EVK of RZ/G3S Linux Start-up Guide`_
```
Comment (kartben) [inline boards/renesas/rzg3s_smarc/doc/index.rst:235]:
```suggestion

.. target-notes::
```

**Review by kartben — CHANGES_REQUESTED** [review]
Some minor fixes needed in the README. Thanks!

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 21]
```diff
@@ -0,0 +1,250 @@
+.. _rzg3s_smarc:
+
+RZ/G3S SMARC Evaluation Board Kit
+#################################
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: Thank you.
I updated as you suggestion.
Please kindly check it.

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 22]
```diff
@@ -0,0 +1,250 @@
...
+  * PMOD x2
+  * USB-Type C for power input
+  * PCIe Gen2 4-lane slot (G3S supports only 1-lane)
+  * M.2 Key E
+  * M.2 Key B and SIM card
+  * Coin cell battery holder (3.0V support)
+
+.. figure:: rzg3s_smarc.webp
+	:align: center
+	:alt: RZ/G3S SMARC Evaluation Board Kit
```
[the remark this answers is not in the document]
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: I updated as you suggestion.

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 20 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst]: I updated as you suggestion.

**Review by kartben — DISMISSED** [review]
+1 for docs, thanks!

**Review by tejlmand — DISMISSED** [review]

**Review by tejlmand — DISMISSED** [review]

**Review by kartben — DISMISSED** [review]

**Comment by kartben** [comment]
@tgorochowik please review.
As a follow-up, it will be nice to update the MAINTAINERS.yml entry for Renesas RZ platform so that it catches all RZ socs/drivers/boards :)

File: boards/renesas/rzg3s_smarc/doc/index.rst [hunk 23]
```diff
@@ -0,0 +1,234 @@
...
+  * M.2 Key E
+  * M.2 Key B and SIM card
+  * Coin cell battery holder (3.0V support)
+
+Hardware
+********
+
+The Renesas RZ/G3S MPU documentation can be found at `RZ/G3S Group Website`_
+
+.. figure:: rzg3s_block_diagram.webp
```
Comment (thaoluonguw) [inline boards/renesas/rzg3s_smarc/doc/index.rst:44]: @binhnguyen2434 : The picture in hardware should be rzg3s_smarc.webp rather than a block diagram.

File: soc/renesas/rz/rzg3s/pinctrl_soc.h [hunk 24]
```diff
@@ -0,0 +1,11 @@
+/*
+ * Copyright (c) 2024 Renesas Electronics Corporation
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#ifndef ZEPHYR_SOC_RENESAS_RZ_RZG3S_PINCTRL_SOC_H_
+#define ZEPHYR_SOC_RENESAS_RZ_RZG3S_PINCTRL_SOC_H_
+
+#include "../common/pinctrl_rzg.h"
```
Comment (thaoluonguw) [inline soc/renesas/rz/rzg3s/pinctrl_soc.h:9]: Should improve for include?

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 23 above)
Comment (binhnguyen2434) [reply boards/renesas/rzg3s_smarc/doc/index.rst:44]: Thank you for the feedback.
Zephyr doc template has updated, [rzg3s_smarc.webp](https://github.com/zephyrproject-rtos/zephyr/pull/82016/files#diff-d55f2e77245a47a9d69c1c61d21192d228e5b7e86070daca6b1d98003d399896) is realized by Zephyr system. No need to mention directly in document.

**Review by nordicjm — DISMISSED** [review]

**Review by tgorochowik — APPROVED** [review]

File: soc/renesas/rz/rzg3s/pinctrl_soc.h (hunk 24 above)
Comment (binhnguyen2434) [reply soc/renesas/rz/rzg3s/pinctrl_soc.h]: Thank you.
I improve this include.
Please kindly check it

File: boards/renesas/rzg3s_smarc/doc/index.rst (hunk 23 above)
Comment (thaoluonguw) [reply boards/renesas/rzg3s_smarc/doc/index.rst:44]: Thanks @binhnguyen2434 for your feedback. I understood.

**Review by thaoluonguw — APPROVED** [review]

**Review by KhiemNguyenT — APPROVED** [review]
LGTM.
Please continue adding more support for other peripherals.

**Review by kartben — APPROVED** [review]
